pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -730,89 +732,127 @@ func (service *SecurityPolicyService) deleteVPCSecurityPolicy(sp types.UID, isGC
 	isDefaultProject := false
 	if vpcInfo.ProjectID == common.DefaultProject {
 		isDefaultProject = true
 	}
 
-	nsxSecurityPolicy.MarkedForDelete = &MarkedForDelete
-
-	// There is no NSX groups/rules in the security policy retrieved from securityPolicy store.
-	// The groups/rules associated the deleting security policy can only be gotten from group/rule store.
-	existingGroups := groupStore.GetByIndex(indexScope, string(sp))
-	service.markDeleteGroups(existingGroups, nsxGroups, sp)
-
-	existingRules := ruleStore.GetByIndex(indexScope, string(sp))
-	service.markDeleteRules(existingRules, nsxRules, sp)
-	nsxSecurityPolicy.Rules = *nsxRules
-
 	var existingNsxShares []*model.Share
 	var existingNsxGroups []*model.Group
 	if isDefaultProject {
-		existingNsxShares = infraShareStore.GetByIndex(indexScope, string(sp))
-		existingNsxGroups = infraGroupStore.GetByIndex(indexScope, string(sp))
+		existingNsxShares = infraShareStore.GetByIndex(indexScope, string(spUID))
+		existingNsxGroups = infraGroupStore.GetByIndex(indexScope, string(spUID))
 	} else {
-		existingNsxShares = projectShareStore.GetByIndex(indexScope, string(sp))
-		existingNsxGroups = projectGroupStore.GetByIndex(indexScope, string(sp))
+		existingNsxShares = projectShareStore.GetByIndex(indexScope, string(spUID))
+		existingNsxGroups = projectGroupStore.GetByIndex(indexScope, string(spUID))
 	}
-	service.markDeleteShares(existingNsxShares, nsxShares, sp)
-	service.markDeleteGroups(existingNsxGroups, nsxShareGroups, sp)
-
-	// wrapHierarchyVpcSecurityPolicy will modify the input security policy, so we need to make a copy for the following store update.
-	finalSecurityPolicyCopy := *nsxSecurityPolicy
-	finalSecurityPolicyCopy.Rules = nsxSecurityPolicy.Rules
+	service.markDeleteShares(existingNsxShares, nsxShares, spUID)
+	service.markDeleteGroups(existingNsxGroups, nsxShareGroups, spUID)
 
 	if !isDefaultProject {
-		_, err = service.manipulateSecurityPolicy(nsxSecurityPolicy, *nsxGroups, *nsxShares, *nsxShareGroups, true, &vpcInfo)
+		if nsxSecurityPolicy != nil {
+			err = service.deleteNSXSecurityPolicy(nsxSecurityPolicy, &vpcInfo)
+			if err != nil {
+				log.Error(err, "Failed to delete NSX SecurityPolicy, rules in VPC", "nsxSecurityPolicyUID", spUID)
+				return err
+			}
+			err = service.applySecurityPolicyStore(nsxSecurityPolicy, *nsxRules, true)
+			if err != nil {
+				return err
+			}
+			log.Info("Successfully deleted NSX SecurityPolicy and rules only", "nsxSecurityPolicyUID", spUID)
+		}
+
+		err = service.deleteNSXSecurityPolicyGroupShare(*nsxGroups, *nsxShares, *nsxShareGroups, &vpcInfo)
+		// Ignore error here to make groups/shares to deleted in GC.
+		// Because NSX SecurityPolicy is deleted, it's unable to get SecurityPolicy from SecurityPolicyStore by requeuing the error.
+		if err != nil {
+			log.Info("NSX SecurityPolicy is deleted, and groups, shares in VPC will be GC", "nsxSecurityPolicyUID", spUID)
```

Comment:
Please polish the sentences including "`, and groups, shares`", they looks not very fluent.